### PR TITLE
Delete Cancer Study By Cascade

### PR DESCRIPTION
**Pull request by angelicaochoa**
# What? Why?
* Added new delete cancer study method that deletes a study by cascade from DB with foreign key constraints enforced.
* Fixed a bug in `DaoCancerStudy.deleteCancerStudy(...)` method - array of SQL statements did not contain clean up statement for `cna_event` or delete `copy_number_seg_file` records referencing the cancer study being deleted

Changes proposed in this pull request:
* Deprecated DaoCancerStudy.deleteCancerStudy(...) methods.
* Fixed a bug in `DaoCancerStudy.deleteCancerStudy(...)`
* Added new method that deletes a study by cascade from DB with foreign key constraints enforced.
* Added method that removes unused references in `cna_event`, `mutation_event`, and `gistic` that may have been unique to a single study that no longer exists in the database.

# Checks
- [ ] Runs on Heroku.
- [x] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [x] Follows the [Google Style Guide](https://github.com/google/styleguide).
- [x] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)
- [x] If this is a feature, the PR is to rc. If this is a bug fix, the PR is to
  hotfix.

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [GifGrabber](http://www.gifgrabber.com/).

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). If you are part of the
cBioPortal organization, notify the approprate team (remove inappropriate):

@n1zea144 
@sheridancbio 
@zheins 

If you are not part of the cBioPortal organization look at who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them here:

Signed-off-by: Angelica Ochoa <aochoa4230@gmail.com>

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 1]
```diff
@@ -449,6 +449,12 @@ public static void deleteAllRecords() throws DaoException {
         }
     }
 
+    /**
+     * 
+     * @param cancerStudyStableId
+     * @throws DaoException 
+     * @deprecated
```
Comment (sheridancbio) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: right

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 2]
```diff
@@ -481,12 +487,49 @@ public static void deleteCancerStudy(String cancerStudyStableId) throws DaoExcep
...
+            deleteCancerStudyByCascade(study.getInternalId());
+        }
+    }
 
     /**
+     * Deletes a cancer study by internal id from db with foreign key constraints enforced.
+     * @param internalCancerStudyId
+     * @throws DaoException 
+     */
+    static void deleteCancerStudyByCascade(int internalCancerStudyId) throws DaoException {
```
Comment (sheridancbio) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: this probably should have a visibility specifier (public or private). Public is probably good.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 3]
```diff
@@ -481,12 +487,49 @@ public static void deleteCancerStudy(String cancerStudyStableId) throws DaoExcep
...
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoCancerStudy.class);
+            pstmt = con.prepareStatement("DELETE FROM cancer_study WHERE cancer_study_id = ?");
+            pstmt.setInt(1, internalCancerStudyId);
+            pstmt.executeUpdate();
```
Comment (sheridancbio) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: I think executeUpdate() is going to return the number of records altered. If this returns zero, then the deletion did not happen, and we probably should throw a DaoException to let the caller know that the call failed.
("Could not find cancer_study_id in table cancer_study")
An interesting question is --- when it succeeds, I wonder if the returned value is 1 (for the one record deleted from cancer_study) or a bigger number (for all the cascaded deletes that happen across all tables). I have no idea what the behavior is, but would guess we will get the bigger number.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 4]
```diff
@@ -481,12 +487,49 @@ public static void deleteCancerStudy(String cancerStudyStableId) throws DaoExcep
...
+            removeCancerStudyFromCache(internalCancerStudyId);
```
Comment (sheridancbio) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: do we need both an explicit removeCancerStudyFromCache(id), and also the more general reCacheAll() below? I'm not familiar with reCacheAll .. but if it deletes the whole cache and then recaches maybe this is not needed here?

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 5]
```diff
@@ -515,6 +558,7 @@ static void deleteCancerStudy(int internalCancerStudyId) throws DaoException {
                 "delete from clinical_patient where INTERNAL_ID IN (select INTERNAL_ID from patient where CANCER_STUDY_ID=?);",
                 "delete from patient where CANCER_STUDY_ID=?;",
                 "delete from copy_number_seg where CANCER_STUDY_ID=?;",
+                "delete from copy_number_seg_file where CANCER_STUDY_ID=?;",
```
Comment (sheridancbio) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: bugfix for old function? good.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java (hunk 4 above)
Comment (zheins) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: I think recacheAll needs to be called to update the other DaoObjects if subsequent operations are going to be performed in this transaction. The front end cache doesn't get updated, but at least this way the other maps will be for any further dao calls.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java (hunk 3 above)
Comment (zheins) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: Checking if this is greater than 0 would be a good idea, but not a huge deal. Small detail.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java (hunk 5 above)
Comment (zheins) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: Are there any other tables that have gotten missed over time? I have a list I can share with you, maybe we should check if any other tables need deleting.

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java [hunk 6]
```diff
@@ -481,14 +487,91 @@ public static void deleteCancerStudy(String cancerStudyStableId) throws DaoExcep
...
+    private static void deleteUnusedReferencesFromDb() throws DaoException {
+        Connection con = null;
+        PreparedStatement pstmt = null;
+        ResultSet rs = null;
+        try {
+            con = JdbcUtil.getDbConnection(DaoCancerStudy.class);
+            
+            // sql statements to remove unused references in mutation_event, cna_event, 
+            // and gistic after a study is deleted. 
+            String[] sqls = {
```
Comment (zheins) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoCancerStudy.java]: I thought we decided to leave this out for now and research a better solution? It's fine to include though.

**Pull request merged**